applets/wncklet/window-list.c

```diff
@@ -139,6 +146,163 @@ static void applet_change_background(MatePanelApplet* applet, MatePanelAppletBac
+
+	if (window == NULL)
+		return NULL;
+
+	scale = gdk_window_get_scale_factor (window);
+	width = gdk_window_get_width (window) * scale;
+	height = gdk_window_get_height (window) * scale;
+
+	/* Generate window screenshot for preview */
+	screenshot = gdk_pixbuf_get_from_window (window, 0, 0, width / scale, height / scale);
+	g_object_unref (window);
+
+	if (screenshot == NULL)
+		return NULL;
+
+	/* Determine whether the contents of the screenshot are empty */
+	pixels = gdk_pixbuf_get_pixels (screenshot);
+	if (!g_strcmp0 ((const char *)pixels, ""))
+	{
+		g_object_unref (screenshot);
+		return NULL;
+	}
+
+	/* Scale to configured size while maintaining aspect ratio */
+	if (width > height)
+	{
+		ratio = (double) height / (double) width;
+		width = MIN(width, tasklist->thumbnail_size);
+		height = width * ratio;
+	}
+	else
+	{
+		ratio = (double) width / (double) height;
+		height = MIN(height, tasklist->thumbnail_size);
+		width = height * ratio;
+	}
+
+	thumbnail = gdk_pixbuf_scale_simple (screenshot, width, height, GDK_INTERP_BILINEAR);
+	g_object_unref (screenshot);
+
+	return thumbnail;
+}
+
+#define PREVIEW_PADDING 5
+static void preview_window_reposition (TasklistData *tasklist, GdkPixbuf *thumbnail)
+{
+	GdkMonitor *monitor;
+	GdkRectangle monitor_geom;
+	int x_pos, y_pos;
+	int width, height;
+
+	width = gdk_pixbuf_get_width (thumbnail);
+	height = gdk_pixbuf_get_height (thumbnail);
+
+	/* Resize window to fit thumbnail */
+	gtk_window_resize (GTK_WINDOW (tasklist->preview), width, height);
+
+	/* Set position at pointer, then re-adjust from there to just outside of the pointer */
+	gtk_window_set_position (GTK_WINDOW (tasklist->preview), GTK_WIN_POS_MOUSE);
+	gtk_window_get_position (GTK_WINDOW (tasklist->preview), &x_pos, &y_pos);
+
+	/* Get geometry of monitor where tasklist is located to calculate correct position of preview */
+	monitor = gdk_display_get_monitor_at_point (gdk_display_get_default (), x_pos, y_pos);
+	gdk_monitor_get_geometry (monitor, &monitor_geom);
+
+	/* Add padding to clear the panel */
+	switch (mate_panel_applet_get_orient (MATE_PANEL_APPLET (tasklist->applet)))
+	{
+		case MATE_PANEL_APPLET_ORIENT_LEFT:
+			x_pos = monitor_geom.width + monitor_geom.x - (width + tasklist->size) - PREVIEW_PADDING;
+			break;
+		case MATE_PANEL_APPLET_ORIENT_RIGHT:
+			x_pos = tasklist->size + PREVIEW_PADDING;
+			break;
+		case MATE_PANEL_APPLET_ORIENT_UP:
+			y_pos = monitor_geom.height + monitor_geom.y - (height + tasklist->size) - PREVIEW_PADDING;
+			break;
+		case MATE_PANEL_APPLET_ORIENT_DOWN:
+		default:
+			y_pos = tasklist->size + PREVIEW_PADDING;
+			break;
+	}
+
+	gtk_window_move (GTK_WINDOW (tasklist->preview), x_pos, y_pos);
+}
+
+static gboolean preview_window_draw (GtkWidget *widget, cairo_t *cr, GdkPixbuf *thumbnail)
+{
+	GtkStyleContext *context;
+
+	context = gtk_widget_get_style_context (widget);
+	gtk_render_icon (context, cr, thumbnail, 0, 0);
+	g_object_unref (thumbnail);
```

Comment:
Why this unref here?

Reply:
Once the image is rendered it's no longer necessary. Would there be a better place to put this? (or is it not needed?)

Reply:
Are you sure that draw signal will always happen once? Anyway, it does not look correct to me.

Reply:
I just checked and it runs only once per hover. However, I'd rather not have something that can cause issues in the future. Should I just remove it or do you have suggestions of where else to put it?

Reply:
>Should I just remove it or do you have suggestions of where else to put it?

No, you should not simply remove it as this memory still needs to be freed.
What you should do is register a [destroy notify when connecting the signal](https://developer.gnome.org/gobject/stable/gobject-Signals.html#g-signal-connect-data) (one line 293):

```c
	g_signal_connect_data (G_OBJECT (tasklist->preview), "draw", G_CALLBACK (preview_window_draw), thumbnail, (GClosureNotify) g_object_unref, 0);
```

Reply:
Thanks!